File: remote-agents/orchestrator/src/lib/git.js
// git.js — thin async wrappers around git CLI for use inside flow scripts.
// Each helper uses runCommand under the hood and surfaces stdout/exit code.
//
// Push is never automatic — the flow author must call push() explicitly.

import { runCommand } from './runCommand.js';

// Get the diff of unstaged + staged changes in the project. Use this to
// build a review prompt for a second agent.
export async function gitDiff({ projectDir, staged = false, paths = [] }) {
  const flag = staged ? '--staged' : '';
  const pathArg = paths.length ? ` -- ${paths.map(quote).join(' ')}` : '';
  const res = await runCommand(`git diff ${flag}${pathArg}`, { cwd: projectDir });
  return res.stdout;
}

// Short-form summary (file + line counts). Cheap and useful for prompts that
// want context without burning tokens on full diffs.
export async function gitDiffStat({ projectDir, staged = false }) {
  const flag = staged ? '--staged' : '';
  const res = await runCommand(`git diff --stat ${flag}`, { cwd: projectDir });
  return res.stdout;
}

// Stage specific files. Avoid `git add -A` — too easy to grab unintended
// files. Always pass the list explicitly.
export async function gitAdd({ projectDir, files }) {
  if (!files || files.length === 0) {
    throw new Error('gitAdd: files list required (no implicit "add all")');
  }
  const quoted = files.map(quote).join(' ');
  const res = await runCommand(`git add ${quoted}`, { cwd: projectDir });
  if (res.exitCode !== 0) {
    throw new Error(`git add failed: ${res.stderr || res.stdout}`);
  }
  return res;
}

// Commit. Builds the message via a temp file so multi-line messages with
// special characters survive intact.
export async function commit({ projectDir, message, files, coAuthor }) {
  if (!message) throw new Error('commit: message is required');

  if (files && files.length > 0) {
    await gitAdd({ projectDir, files });
  }

  const fullMessage = coAuthor
    ? `${message.trim()}\n\nCo-Authored-By: ${coAuthor} <noreply@anthropic.com>`
    : message.trim();

  // Use -F - to read message from stdin, sidestepping all shell escaping
  const res = await runCommand(`git commit -F -`, {
    cwd: projectDir,
    input: fullMessage,
  });

  if (res.exitCode !== 0) {
    throw new Error(`git commit failed: ${res.stderr || res.stdout}`);
  }
  return res;
}

// Push. Explicit, never inferred. Pass remote/branch if you want non-defaults.
export async function push({ projectDir, remote = 'origin', branch = null, force = false } = {}) {
  if (force && (branch === 'main' || branch === 'master')) {
    throw new Error(`push: refusing to force-push to ${branch} via this primitive`);
  }
  const parts = ['git push'];
  if (force) parts.push('--force-with-lease');
  parts.push(remote);
  if (branch) parts.push(branch);

  const res = await runCommand(parts.join(' '), { cwd: projectDir });
  if (res.exitCode !== 0) {
    throw new Error(`git push failed: ${res.stderr || res.stdout}`);
  }
  return res;
}

// Current branch — useful when the flow wants to push to whatever we're on.
export async function currentBranch({ projectDir }) {
  const res = await runCommand('git rev-parse --abbrev-ref HEAD', { cwd: projectDir });
  return res.stdout.trim();
}

// Cheap "is the tree dirty?" check.
export async function isDirty({ projectDir }) {
  const res = await runCommand('git status --porcelain', { cwd: projectDir });
  return res.stdout.trim().length > 0;
}

function quote(s) {
  // Minimal POSIX-safe single-quote escape; works under cmd.exe too because
  // we pass via shell: true.
  if (/^[A-Za-z0-9_./-]+$/.test(s)) return s;
  return `"${String(s).replace(/"/g, '\\"')}"`;
}

File: src/RemoteAgents/Actors/Git/Git.cs
using RemoteAgents.Flows;
using RemoteAgents.Tools.CommandLine;

namespace RemoteAgents.Actors.Git;

public sealed class Git(string projectDir)
{
    public Flow.Operation<DirtyArgs, DirtyResult> CheckDirty { get; } = new DirtyOp(projectDir);
    public Flow.Operation<DiffArgs, DiffResult> Diff { get; } = new DiffOp(projectDir);
    public Flow.Operation<ChangedFilesArgs, ChangedFilesResult> ChangedFiles { get; } = new ChangedFilesOp(projectDir);
    public Flow.Operation<CommitArgs, GitCommitResult> Commit { get; } = new CommitOp(projectDir);
    public Flow.Operation<PushArgs, GitPushResult> Push { get; } = new PushOp(projectDir);

    private sealed class DirtyOp(string dir) : Flow.Operation<DirtyArgs, DirtyResult>
    {
        protected override async Task<DirtyResult> Invoke(DirtyArgs args, CancellationToken ct)
        {
            var res = (await RunCommand.RunAsync("git status --porcelain", new RunCommandOptions(Cwd: dir), ct))
                .EnsureOk("git status");
            return new DirtyResult(res.Stdout.Trim().Length > 0);
        }
    }

    private sealed class DiffOp(string dir) : Flow.Operation<DiffArgs, DiffResult>
    {
        protected override async Task<DiffResult> Invoke(DiffArgs args, CancellationToken ct)
        {
            var res = (await RunCommand.RunAsync("git diff", new RunCommandOptions(Cwd: dir), ct))
                .EnsureOk("git diff");
            return new DiffResult(res.Stdout, CountFiles(res.Stdout));
        }
    }

    private sealed class ChangedFilesOp(string dir) : Flow.Operation<ChangedFilesArgs, ChangedFilesResult>
    {
        protected override async Task<ChangedFilesResult> Invoke(ChangedFilesArgs args, CancellationToken ct)
        {
            var res = (await RunCommand.RunAsync("git status --porcelain", new RunCommandOptions(Cwd: dir), ct))
                .EnsureOk("git status");
            return new ChangedFilesResult(ParsePorcelainPaths(res.Stdout));
        }
    }

    private sealed class CommitOp(string dir) : Flow.Operation<CommitArgs, GitCommitResult>
    {
        protected override async Task<GitCommitResult> Invoke(CommitArgs args, CancellationToken ct)
        {
            if (string.IsNullOrWhiteSpace(args.Message))
                throw new ArgumentException("git commit: message is required", nameof(args));
            if (args.Files.Count == 0)
                throw new ArgumentException("git commit: an explicit file list is required (no implicit add -A)", nameof(args));

            var quoted = string.Join(' ', args.Files.Select(Shell.QuoteArg));
            (await RunCommand.RunAsync($"git add {quoted}", new RunCommandOptions(Cwd: dir), ct)).EnsureOk("git add");

            var fullMessage = args.CoAuthor is null
                ? args.Message.Trim()
                : $"{args.Message.Trim()}\n\nCo-Authored-By: {args.CoAuthor} <noreply@anthropic.com>";
            (await RunCommand.RunAsync("git commit -F -", new RunCommandOptions(Cwd: dir, Input: fullMessage), ct))
                .EnsureOk("git commit");

            var head = (await RunCommand.RunAsync("git rev-parse HEAD", new RunCommandOptions(Cwd: dir), ct))
                .EnsureOk("git rev-parse");
            return new GitCommitResult(head.Stdout.Trim(), FirstLine(args.Message));
        }
    }

    private sealed class PushOp(string dir) : Flow.Operation<PushArgs, GitPushResult>
    {
        protected override async Task<GitPushResult> Invoke(PushArgs args, CancellationToken ct)
        {
            if (args.Force && IsProtected(args.Branch))
                throw new InvalidOperationException($"git push: refusing to force-push to {args.Branch}");

            var target = args.Branch ?? await CurrentBranchAsync(dir, ct);
            if (args.Force && IsProtected(target))
                throw new InvalidOperationException($"git push: refusing to force-push to {target}");

            var parts = new List<string> { "git push" };
            if (args.Force) parts.Add("--force-with-lease");
            parts.Add(args.Remote);
            parts.Add(target);
            (await RunCommand.RunAsync(string.Join(' ', parts), new RunCommandOptions(Cwd: dir), ct)).EnsureOk("git push");
            return new GitPushResult(args.Remote, target);
        }
    }

    private static async Task<string> CurrentBranchAsync(string projectDir, CancellationToken ct)
    {
        var res = (await RunCommand.RunAsync("git rev-parse --abbrev-ref HEAD", new RunCommandOptions(Cwd: projectDir), ct))
            .EnsureOk("git rev-parse");
        return res.Stdout.Trim();
    }

    private static bool IsProtected(string? branch) => branch is "main" or "master";

    private static IReadOnlyList<string> ParsePorcelainPaths(string stdout)
    {
        var paths = new List<string>();
        foreach (var raw in stdout.Split('\n'))
        {
            var line = raw.TrimEnd('\r');
            if (line.Length < 4) continue;
            var rest = line[3..];
            var arrow = rest.IndexOf(" -> ", StringComparison.Ordinal);
            var path = (arrow >= 0 ? rest[(arrow + 4)..] : rest).Trim().Trim('"');
            if (path.Length > 0) paths.Add(path);
        }
        return paths;
    }

    private static int CountFiles(string diff)
    {
        var count = 0;
        foreach (var line in diff.Split('\n'))
            if (line.StartsWith("diff --git ", StringComparison.Ordinal)) count++;
        return count;
    }

    private static string FirstLine(string message)
    {
        var nl = message.IndexOf('\n');
        return (nl >= 0 ? message[..nl] : message).Trim();
    }
}
